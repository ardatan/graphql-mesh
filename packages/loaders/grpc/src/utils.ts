import { existsSync } from "fs";
import type { SchemaComposer } from "graphql-compose";
import { isAbsolute, join } from "path";
import type { INamespace, Root } from "protobufjs";
import { getGraphQLScalarForGrpc, isGrpcScalar } from "./scalars.js";
import lodashHas from 'lodash.has';

export function addIncludePathResolver(root: Root, includePaths: string[]): void {
  const originalResolvePath = root.resolvePath;
  root.resolvePath = (origin: string, target: string) => {
    if (isAbsolute(target)) {
      return target;
    }
    for (const directory of includePaths) {
      const fullPath: string = join(directory, target);
      if (existsSync(fullPath)) {
        return fullPath;
      }
    }
    const path = originalResolvePath(origin, target);
    if (path === null) {
      console.warn(`${target} not found in any of the include paths ${includePaths}`);
    }
    return path;
  };
}

export function getTypeName(
  schemaComposer: SchemaComposer,
  pathWithName: string[] | undefined,
  isInput: boolean,
) {
  if (pathWithName?.length) {
    const baseTypeName = pathWithName.filter(Boolean).join('__');
    if (isGrpcScalar(baseTypeName)) {
      return getGraphQLScalarForGrpc(baseTypeName);
    }
    if (schemaComposer.isEnumType(baseTypeName)) {
      return baseTypeName;
    }
    return isInput ? baseTypeName + '_Input' : baseTypeName;
  }
  return 'Void';
}

export function walkToFindTypePath(
  rootJson: INamespace,
  pathWithName: string[],
  baseTypePath: string[],
) {
  const currentWalkingPath = [...pathWithName];
  while (!lodashHas(rootJson.nested, currentWalkingPath.concat(baseTypePath).join('.nested.'))) {
    if (!currentWalkingPath.length) {
      break;
    }
    currentWalkingPath.pop();
  }
  return currentWalkingPath.concat(baseTypePath);
}
