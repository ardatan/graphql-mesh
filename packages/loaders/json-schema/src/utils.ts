import { OperationTypeNode } from 'graphql';
import { JSONSchemaOperationConfig, JSONSchemaPubSubOperationConfig, HTTPMethod } from './types';

export function isPubSubOperationConfig(
  operationConfig: JSONSchemaOperationConfig
): operationConfig is JSONSchemaPubSubOperationConfig {
  return 'pubsubTopic' in operationConfig;
}

export function getOperationMetadata(operationConfig: JSONSchemaOperationConfig) {
  let httpMethod: HTTPMethod;
  let operationType: OperationTypeNode;
  let rootTypeName: 'Query' | 'Mutation' | 'Subscription';
  if (isPubSubOperationConfig(operationConfig)) {
    httpMethod = null;
    operationType = 'subscription' as OperationTypeNode;
    rootTypeName = 'Subscription';
  } else {
    httpMethod = operationConfig.method;
    // Fix compability with Mesh handler
    operationType = operationConfig.type.toLowerCase() as OperationTypeNode;
    if (!httpMethod) {
      if (operationType === 'mutation') {
        httpMethod = 'POST';
      } else {
        httpMethod = 'GET';
      }
    }
    if (!rootTypeName) {
      if (httpMethod === 'GET') {
        rootTypeName = 'Query';
      }
    }
    rootTypeName = operationType === 'query' ? 'Query' : 'Mutation';
  }
  return {
    httpMethod,
    operationType,
    rootTypeName,
    fieldName: operationConfig.field,
  };
}
