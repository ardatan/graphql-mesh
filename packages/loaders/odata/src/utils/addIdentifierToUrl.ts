export function addIdentifierToUrl(
  url: URL,
  identifierFieldName: string,
  identifierFieldTypeRef: string,
  args: any,
) {
  url.href += `/${args[identifierFieldName]}/`;
}
