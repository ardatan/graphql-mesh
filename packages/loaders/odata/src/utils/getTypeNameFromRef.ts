import { pascalCase } from 'pascal-case';

interface GetNamespaceFromTypeRefOpts {
  namespaces: Set<string>;
  typeRef: string;
}

function getNamespaceFromTypeRef({ typeRef, namespaces }: GetNamespaceFromTypeRefOpts) {
  let namespace = '';
  namespaces?.forEach(el => {
    if (
      typeRef.startsWith(el) &&
      el.length > namespace.length && // It can be deeper namespace
      !typeRef.replace(el + '.', '').includes('.') // Typename cannot have `.`
    ) {
      namespace = el;
    }
  });
  return namespace;
}

const SCALARS = new Map<string, string>([
  ['Edm.Binary', 'String'],
  ['Edm.Stream', 'String'],
  ['Edm.String', 'String'],
  ['Edm.Int16', 'Int'],
  ['Edm.Byte', 'Byte'],
  ['Edm.Int32', 'Int'],
  ['Edm.Int64', 'BigInt'],
  ['Edm.Double', 'Float'],
  ['Edm.Boolean', 'Boolean'],
  ['Edm.Guid', 'GUID'],
  ['Edm.DateTimeOffset', 'DateTime'],
  ['Edm.Date', 'Date'],
  ['Edm.TimeOfDay', 'String'],
  ['Edm.Single', 'Float'],
  ['Edm.Duration', 'ISO8601Duration'],
  ['Edm.Decimal', 'Float'],
  ['Edm.SByte', 'Byte'],
  ['Edm.GeographyPoint', 'String'],
]);

export function getTypeNameFromRef({
  typeRef,
  isInput,
  isRequired,
  aliasNamespaceMap,
  multipleSchemas,
  namespaces,
  isEnumType,
}: {
  typeRef: string;
  isInput: boolean;
  isRequired: boolean;
  aliasNamespaceMap: Map<string, string>;
  multipleSchemas: boolean;
  namespaces: Set<string>;
  isEnumType(typeName: string): boolean;
}) {
  const typeRefArr = typeRef.split('Collection(');
  const arrayDepth = typeRefArr.length;
  let actualTypeRef = typeRefArr.join('').split(')').join('');
  const typeNamespace = getNamespaceFromTypeRef({
    typeRef: actualTypeRef,
    namespaces,
  });
  if (aliasNamespaceMap.has(typeNamespace)) {
    const alias = aliasNamespaceMap.get(typeNamespace);
    actualTypeRef = actualTypeRef.replace(typeNamespace, alias);
  }
  const actualTypeRefArr = actualTypeRef.split('.');
  const typeName = multipleSchemas
    ? pascalCase(actualTypeRefArr.join('_'))
    : actualTypeRefArr[actualTypeRefArr.length - 1];
  let realTypeName = typeName;
  if (SCALARS.has(actualTypeRef)) {
    realTypeName = SCALARS.get(actualTypeRef);
  } else if (isEnumType(typeName)) {
    realTypeName = typeName;
  } else if (isInput) {
    realTypeName += 'Input';
  }
  const fakeEmptyArr = new Array(arrayDepth);
  realTypeName = fakeEmptyArr.join('[') + realTypeName + fakeEmptyArr.join(']');
  if (isRequired) {
    realTypeName += '!';
  }
  return realTypeName;
}
