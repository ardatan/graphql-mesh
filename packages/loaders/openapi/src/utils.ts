import { camelCase } from 'change-case';

export function getFieldNameFromPath(path: string, method: string, responseTypeSchemaRef: string) {
  // Replace identifiers with "by"
  path = path.split('{').join('by_').split('}').join('');

  const [actualPartsStr, allQueryPartsStr] = path.split('?');

  const actualParts = actualPartsStr.split('/').filter(Boolean);

  let fieldNameWithoutMethod = actualParts.join('_');

  // If path doesn't give any field name without identifiers, we can use the return type with HTTP Method name
  if ((!fieldNameWithoutMethod || fieldNameWithoutMethod.startsWith('by')) && responseTypeSchemaRef) {
    const refArr = responseTypeSchemaRef.split('/');
    // lowercase looks better in the schema
    const prefix = camelCase(refArr[refArr.length - 1]);
    if (fieldNameWithoutMethod) {
      fieldNameWithoutMethod = prefix + '_' + fieldNameWithoutMethod;
    } else {
      fieldNameWithoutMethod = prefix;
    }
  }

  if (allQueryPartsStr) {
    const queryParts = allQueryPartsStr.split('&');
    for (const queryPart of queryParts) {
      const [queryName] = queryPart.split('=');
      fieldNameWithoutMethod += '_' + 'by' + '_' + queryName;
    }
  }

  // get_ doesn't look good in field names
  if (method === 'get') {
    return fieldNameWithoutMethod;
  }
  return method + '_' + fieldNameWithoutMethod;
}
