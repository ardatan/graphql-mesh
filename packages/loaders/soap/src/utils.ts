import { X2jOptions } from 'fast-xml-parser';

export interface SoapAnnotations {
  elementName: string;
  bindingNamespace: string;
  endpoint: string;
}

export const PARSE_XML_OPTIONS: Partial<X2jOptions> = {
  attributeNamePrefix: '',
  attrNodeName: 'attributes',
  textNodeName: 'innerText',
  ignoreAttributes: false,
  ignoreNameSpace: true,
  arrayMode: true,
  allowBooleanAttributes: true,
};
