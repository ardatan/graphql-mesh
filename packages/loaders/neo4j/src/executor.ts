import { DirectiveNode, GraphQLSchema, visit } from 'graphql';
import { Driver } from 'neo4j-driver';
import { Logger, MeshPubSub } from '@graphql-mesh/types';
import { createDefaultExecutor } from '@graphql-tools/delegate';
import { Executor, getDirective, getDocumentNodeFromSchema } from '@graphql-tools/utils';
import { getDriverFromOpts } from './driver.js';
import { getExecutableSchemaFromTypeDefsAndDriver } from './schema.js';

export interface Neo4JExecutorOpts {
  schema: GraphQLSchema;
  driver?: Driver;
  pubsub?: MeshPubSub;
  logger?: Logger;
}

function filterIntrospectionDefinitions<TASTNode extends { directives?: readonly DirectiveNode[] }>(
  node: TASTNode,
): TASTNode | null {
  if (!node.directives?.some(directive => directive.name.value === 'introspection')) {
    return null;
  }
  return node;
}

export async function getNeo4JExecutor(opts: Neo4JExecutorOpts): Promise<Executor> {
  const transportDirectives = getDirective(opts.schema, opts.schema, 'transport');
  if (!transportDirectives?.length) {
    throw new Error('No transport directive found on the schema!');
  }
  const {
    location: endpoint,
    options: { database, auth },
  } = transportDirectives[0];
  let driver = opts.driver;
  if (!driver) {
    driver = getDriverFromOpts({
      endpoint,
      auth,
      logger: opts.logger,
    });
  }
  let typeDefs = getDocumentNodeFromSchema(opts.schema);
  const astVisitor = {
    enter: filterIntrospectionDefinitions,
  };
  typeDefs = visit(typeDefs, {
    EnumTypeDefinition: astVisitor,
    ObjectTypeDefinition: astVisitor,
    InterfaceTypeDefinition: astVisitor,
    UnionTypeDefinition: astVisitor,
    InputObjectTypeDefinition: astVisitor,
    FieldDefinition: astVisitor,
    // DirectiveDefinition: astVisitor,
    ScalarTypeDefinition: astVisitor,
    EnumValueDefinition: astVisitor,
    InputValueDefinition: astVisitor,
  });
  const executableSchema = await getExecutableSchemaFromTypeDefsAndDriver({
    driver,
    logger: opts.logger,
    pubsub: opts.pubsub,
    typeDefs,
  });
  const defaultExecutor = createDefaultExecutor(executableSchema);
  const sessionConfig = {
    database,
  };

  return function neo4JExecutor(args) {
    return defaultExecutor({
      ...args,
      context: {
        ...args.context,
        sessionConfig,
      },
    });
  };
}
