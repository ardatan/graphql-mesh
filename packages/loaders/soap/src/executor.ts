import {
  execute,
  GraphQLFieldResolver,
  GraphQLOutputType,
  GraphQLResolveInfo,
  GraphQLSchema,
  isListType,
  isNonNullType,
} from 'graphql';
import { parse as parseXML, j2xParser as JSONToXMLConverter } from 'fast-xml-parser';
import { MeshFetch } from '@graphql-mesh/types';
import { PARSE_XML_OPTIONS, SoapAnnotations } from './utils.js';
import { Executor, getDirective, getRootTypes } from '@graphql-tools/utils';

function isOriginallyListType(type: GraphQLOutputType): boolean {
  if (isNonNullType(type)) {
    return isOriginallyListType(type.ofType);
  }
  return isListType(type);
}

const defaultFieldResolver: GraphQLFieldResolver<any, any> = function soapDefaultResolver(root, args, context, info) {
  const rootField = root[info.fieldName];
  if (typeof rootField === 'function') {
    return rootField(args, context, info);
  }
  const fieldValue = rootField;
  const isArray = Array.isArray(fieldValue);
  const isPlural = isOriginallyListType(info.returnType);
  if (isPlural && !isArray) {
    return [fieldValue];
  }
  if (!isPlural && isArray) {
    return fieldValue[0];
  }
  return fieldValue;
};

function normalizeArgsForConverter(args: any): any {
  if (args != null) {
    if (typeof args === 'object') {
      for (const key in args) {
        args[key] = normalizeArgsForConverter(args[key]);
      }
    } else {
      return {
        innerText: args,
      };
    }
  }
  return args;
}

function normalizeResult(result: any) {
  if (result != null && typeof result === 'object') {
    for (const key in result) {
      if (key === 'innerText') {
        return result.innerText;
      }
      result[key] = normalizeResult(result[key]);
    }
    if (Array.isArray(result) && result.length === 1) {
      return result[0];
    }
  }
  return result;
}

type RootValueMethod = (args: any, context: any, info: GraphQLResolveInfo) => Promise<any>;

function createRootValueMethod(soapAnnotations: SoapAnnotations, fetchFn: MeshFetch): RootValueMethod {
  const jsonToXMLConverter = new JSONToXMLConverter({
    attributeNamePrefix: '',
    attrNodeName: 'attributes',
    textNodeName: 'innerText',
  });

  return async function rootValueMethod(args: any, context: any, info: GraphQLResolveInfo) {
    const requestJson = {
      'soap:Envelope': {
        attributes: {
          'xmlns:soap': 'http://www.w3.org/2003/05/soap-envelope',
        },
        'soap:Body': {
          attributes: {
            xmlns: soapAnnotations.bindingNamespace,
          },
          ...normalizeArgsForConverter(args),
        },
      },
    };
    const requestXML = jsonToXMLConverter.parse(requestJson);
    const response = await fetchFn(
      soapAnnotations.baseUrl,
      {
        method: 'POST',
        body: requestXML,
        headers: {
          'Content-Type': 'application/soap+xml; charset=utf-8',
        },
      },
      context,
      info
    );
    const responseXML = await response.text();
    const responseJSON = parseXML(responseXML, PARSE_XML_OPTIONS);
    return normalizeResult(responseJSON.Envelope[0].Body[0][soapAnnotations.elementName]);
  };
}

function createRootValue(schema: GraphQLSchema, fetchFn: MeshFetch) {
  const rootValue: Record<string, RootValueMethod> = {};
  const rootTypes = getRootTypes(schema);
  for (const rootType of rootTypes) {
    const rootFieldMap = rootType.getFields();
    for (const fieldName in rootFieldMap) {
      const annotations = getDirective(schema, rootFieldMap[fieldName], 'soap');
      const soapAnnotations: SoapAnnotations = Object.assign({}, ...annotations);
      rootValue[fieldName] = createRootValueMethod(soapAnnotations, fetchFn);
    }
  }
  return rootValue;
}

export function createExecutorFromSchemaAST(schema: GraphQLSchema, fetchFn: MeshFetch) {
  let rootValue: Record<string, RootValueMethod>;
  return function soapExecutor({ document, variables, context }) {
    if (!rootValue) {
      rootValue = createRootValue(schema, fetchFn);
    }
    return execute({
      schema,
      document,
      rootValue,
      contextValue: context,
      variableValues: variables,
      fieldResolver: defaultFieldResolver,
    });
  } as Executor;
}
