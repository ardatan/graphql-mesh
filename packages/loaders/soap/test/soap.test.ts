import { parse } from 'graphql';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { fetch } from '@whatwg-node/fetch';
import { createExecutorFromSchemaAST } from '../src/executor.js';
import { SOAPLoader } from '../src/SOAPLoader.js';

describe('SOAP Loader', () => {
  it('should generate the schema correctly', async () => {
    const soapLoader = new SOAPLoader({
      fetch,
    });
    await soapLoader.fetchWSDL('https://www.w3schools.com/xml/tempconvert.asmx?WSDL');
    const schema = soapLoader.buildSchema();
    expect(printSchemaWithDirectives(schema)).toMatchSnapshot();
  });
  it('should execute SOAP calls correctly', async () => {
    const soapLoader = new SOAPLoader({
      fetch,
    });
    await soapLoader.fetchWSDL('https://www.crcind.com/csp/samples/SOAP.Demo.cls?WSDL');
    const schema = soapLoader.buildSchema();
    const executor = createExecutorFromSchemaAST(schema, fetch);
    const result: any = await executor({
      document: parse(/* GraphQL */ `
        mutation AddInteger {
          s0_SOAPDemo_SOAPDemoSoap_AddInteger(AddInteger: { Arg1: 2, Arg2: 3 }) {
            AddIntegerResult
          }
        }
      `),
    });
    // eslint-disable-next-line eqeqeq
    expect(result?.data?.s0_SOAPDemo_SOAPDemoSoap_AddInteger.AddIntegerResult == 5).toBeTruthy();
  });
});
