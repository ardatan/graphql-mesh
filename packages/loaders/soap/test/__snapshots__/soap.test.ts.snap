// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SOAP Loader should generate the schema correctly 1`] = `
"schema {
  query: Query
  mutation: Mutation
}

directive @soap(elementName: String, bindingNamespace: String, endpoint: String) on FIELD_DEFINITION

type Query {
  placeholder: Void
}

"Represents NULL values"
scalar Void

type Mutation {
  tns_TempConvert_TempConvertSoap_FahrenheitToCelsius(FahrenheitToCelsius: tns_FahrenheitToCelsius_Input): tns_FahrenheitToCelsiusResponse @soap(elementName: "FahrenheitToCelsiusResponse", bindingNamespace: "https://www.w3schools.com/xml/", endpoint: "http://www.w3schools.com/xml/tempconvert.asmx")
  tns_TempConvert_TempConvertSoap_CelsiusToFahrenheit(CelsiusToFahrenheit: tns_CelsiusToFahrenheit_Input): tns_CelsiusToFahrenheitResponse @soap(elementName: "CelsiusToFahrenheitResponse", bindingNamespace: "https://www.w3schools.com/xml/", endpoint: "http://www.w3schools.com/xml/tempconvert.asmx")
  tns_TempConvert_TempConvertSoap12_FahrenheitToCelsius(FahrenheitToCelsius: tns_FahrenheitToCelsius_Input): tns_FahrenheitToCelsiusResponse @soap(elementName: "FahrenheitToCelsiusResponse", bindingNamespace: "https://www.w3schools.com/xml/", endpoint: "http://www.w3schools.com/xml/tempconvert.asmx")
  tns_TempConvert_TempConvertSoap12_CelsiusToFahrenheit(CelsiusToFahrenheit: tns_CelsiusToFahrenheit_Input): tns_CelsiusToFahrenheitResponse @soap(elementName: "CelsiusToFahrenheitResponse", bindingNamespace: "https://www.w3schools.com/xml/", endpoint: "http://www.w3schools.com/xml/tempconvert.asmx")
  tns_TempConvert_TempConvertHttpPost_FahrenheitToCelsius(Fahrenheit: String = ""): String @soap(elementName: "string", bindingNamespace: "https://www.w3schools.com/xml/", endpoint: "http://www.w3schools.com/xml/tempconvert.asmx")
  tns_TempConvert_TempConvertHttpPost_CelsiusToFahrenheit(Celsius: String = ""): String @soap(elementName: "string", bindingNamespace: "https://www.w3schools.com/xml/", endpoint: "http://www.w3schools.com/xml/tempconvert.asmx")
}

type tns_FahrenheitToCelsiusResponse {
  FahrenheitToCelsiusResult: String!
}

input tns_FahrenheitToCelsius_Input {
  Fahrenheit: String
}

type tns_CelsiusToFahrenheitResponse {
  CelsiusToFahrenheitResult: String!
}

input tns_CelsiusToFahrenheit_Input {
  Celsius: String
}"
`;
